precompiles/unified-accounts/src/lib.rs

```diff
@@ -0,0 +1,119 @@
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use astar_primitives::evm::{UnifiedAddress, UnifiedAddressMapper};
+use core::marker::PhantomData;
+use fp_evm::Precompile;
+use fp_evm::{PrecompileHandle, PrecompileOutput};
+use frame_support::dispatch::Dispatchable;
+use precompile_utils::{
+    succeed, Address, EvmDataWriter, EvmResult, FunctionModifier, PrecompileHandleExt,
+};
+use sp_core::{crypto::AccountId32, H256};
+use sp_std::prelude::*;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[precompile_utils::generate_function_selector]
+#[derive(Debug, PartialEq)]
+pub enum Action {
+    GetEvmAddressOrDefault = "get_evm_address_or_default(bytes32)",
+    GetNativeAddressOrDefault = "get_native_address_or_default(address)",
+}
+
+/// A precompile that expose AU related functions.
+pub struct UnifiedAccountsPrecompile<T, UA>(PhantomData<(T, UA)>);
+
+impl<R, UA> Precompile for UnifiedAccountsPrecompile<R, UA>
+where
+    R: pallet_evm::Config + pallet_unified_accounts::Config,
+    <<R as frame_system::Config>::RuntimeCall as Dispatchable>::RuntimeOrigin:
+        From<Option<R::AccountId>>,
+    <R as frame_system::Config>::AccountId: From<AccountId32> + Into<AccountId32>,
+    UA: UnifiedAddressMapper<R::AccountId>,
+{
+    fn execute(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+        log::trace!(target: "au-precompile", "Execute input = {:?}", handle.input());
+
+        let selector = handle.read_selector()?;
+
+        handle.check_function_modifier(FunctionModifier::View)?;
+
+        match selector {
+            // Dispatchables
+            Action::GetEvmAddressOrDefault => Self::get_evm_address_or_default(handle),
+            Action::GetNativeAddressOrDefault => Self::get_native_address_or_default(handle),
+        }
+    }
+}
+
+impl<R, UA> UnifiedAccountsPrecompile<R, UA>
+where
+    R: pallet_evm::Config + pallet_unified_accounts::Config,
+    <<R as frame_system::Config>::RuntimeCall as Dispatchable>::RuntimeOrigin:
+        From<Option<R::AccountId>>,
+    <R as frame_system::Config>::AccountId: From<AccountId32> + Into<AccountId32>,
+    UA: UnifiedAddressMapper<R::AccountId>,
+{
+    fn get_evm_address_or_default(
+        handle: &mut impl PrecompileHandle,
+    ) -> EvmResult<PrecompileOutput> {
+        let mut input = handle.read_input()?;
+        input.expect_arguments(1)?;
+        let account_id = input.read::<H256>()?;
+        log::trace!(target: "au-precompile", "get_evm_address_or_default account_id (Bytes) : {:?}",account_id);
+        let res: (Address, bool) =
+            match UA::to_h160_or_default(&AccountId32::new(account_id.0).into()) {
+                UnifiedAddress::Mapped(address) => (address.into(), true),
+                UnifiedAddress::Default(address) => (address.into(), false),
+            };
+        log::trace!(target: "au-precompile", "accountId : {:?}, (Address,bool): {:?}",account_id, res);
```

Comment:
```suggestion
        let account_id = AccountId32::new(input.read::<H256>()?.into()).into() ;
        log::trace!(target: "au-precompile", "get_evm_address_or_default account_id (Bytes) : {:?}",account_id);
        let res: (Address, bool) = match UA::to_h160_or_default(&account_id) {
            UnifiedAddress::Mapped(address) => (address.into(), true),
            UnifiedAddress::Default(address) => (address.into(), false),
        };
        log::trace!(target: "au-precompile", "accountId : {:?}, (Address,bool): {:?}",account_id, res);
```

Reply:
Better to use `into()` impl than accessing inner wrapped bytes from `H256`

Reply:
makes sense 👍🏻